kernel/vmbytegen.ml

```diff
@@ -831,9 +834,36 @@ and compile_instance env cenv u sz cont =
     | None -> compile_structured_constant cenv (Const_quality q) sz cont
     | Some idx -> pos_instance cenv sz :: Kfield 0 :: Kfield idx :: cont
     in
-    let comp_univ cenv l sz cont = match Univ.Level.var_index l with
-    | None -> compile_structured_constant cenv (Const_univ_level l) sz cont
-    | Some idx -> pos_instance cenv sz :: Kfield 1 :: Kfield idx :: cont
+    let comp_univ cenv u sz cont = 
+      let () = set_max_stack_size cenv sz in
+      (* Length is always >= 1 *)
+      if is_closed_univ u then
+        compile_structured_constant cenv (Const_univ u) sz cont
+      else
+        let u = Univ.Universe.repr u in
+        let len = List.length u in
+        let () = set_max_stack_size cenv (sz + len * 2 - 1) in (* u1 :: ... :: un :: l1 :: ... :: ln *)
+        let rec push_univs (contl, contu) n = function
+          | [] -> contl @ contu
+          | (l, k as x) :: xs ->
+            match Univ.Level.var_index l with
+            | None -> (* Closed level *)
+              let contl = Kconst (Const_b0 0) :: Kpush :: contl in
+              let contu = Kconst (Const_univ (Univ.Universe.of_expr x)) ::
+                (if Int.equal n 0 then contu else Kpush :: contu) in
+              push_univs (contl, contu) (succ n) xs
+            | Some idx ->
+              let contl = Kconst (Const_b0 k) :: Kpush :: contl in
+              let contu =
+                  pos_instance cenv (sz + len (* for all increments *) + (len - n - 1) (* for previous universes *))
+                  :: Kfield 1 :: Kfield idx :: (if Int.equal n 0 then contu else Kpush :: contu) in (* We keep the first universe in accu *)
+              push_univs (contl, contu) (succ n) xs
+        in
+        (* Kmakeuniv n (n always > 0) expects the first universe [u0] in accumulator,
+           then a stack of shape [u1 :: .. :: un :: i0 :: .. :: in], where [ui]'s are universes
+           and [ik]'s are constant increments. It produces a valid universe representing the
+           supremum of [u0+i0 ... un+in]  *)
+        push_univs ([], Kmakeuniv len :: cont) 0 u
```

Comment:
the approach at https://github.com/coq/coq/pull/18968 seems to have much simpler code and does 1 callback / instance with variables instead of 1 callback / universe in the instance with variables so may be faster too.